apps/src/music/blockly/MusicBlocklyWorkspace.js

```diff
@@ -211,22 +215,37 @@ export default class MusicBlocklyWorkspace {
         ].includes(block.type)
       ) {
         const id = block.getFieldValue(TRIGGER_FIELD);
-        events[this.triggerIdToEvent(id)] = {
+        this.compiledEvents[this.triggerIdToEvent(id)] = {
           code:
             Blockly.JavaScript.blockToCode(block) + functionImplementationsCode
         };
       }
     });
 
-    this.codeHooks = {};
+    const currentEventsJson = JSON.stringify(this.compiledEvents);
+    const lastExecutedEventsJson = JSON.stringify(this.lastExecutedEvents);
 
-    console.log('executeSong', events);
+    if (currentEventsJson === lastExecutedEventsJson) {
+      console.log("Code hasn't changed since last execute.");
+      return false;
+    }
 
-    CustomMarshalingInterpreter.evalWithEvents(scope, events).hooks.forEach(
-      hook => {
-        this.codeHooks[hook.name] = hook.func;
-      }
-    );
+    console.log('Compiled song.', this.compiledEvents);
+
+    return true;
+  }
+
+  executeCompiledSong(scope) {
+    console.log('Executing compiled song.');
```

Comment:
Should we add a check here in case executeCompiledSong is called before compileSong?

Reply:
Good idea; added a warning and an early-out.